File: mergesort.rs
fn mergesort(arr: &mut [i32]) {
  if arr.len() == 1 { return; }
  if arr.len() == 2 {
    if arr[0] > arr[1] {
      arr[0] ^= arr[1];
      arr[1] ^= arr[0];
      arr[0] ^= arr[1];
    }
  }
  let p = arr.len() / 2;
  mergesort(&mut arr[0 .. p]);
  mergesort(&mut arr[p ..]);

  let mut arr2 = vec![0; arr.len()];
  let mut p1 = 0;
  let mut p2 = p;
  let mut f = 0;

  loop {
    if p1 < p && p2 < arr.len() {
      if arr[p1] <= arr[p2] {
        arr2[f] = arr[p1];
        p1 += 1;
      } else {
        arr2[f] = arr[p2];
        p2 += 1;
      }
    } else if p1 < p {
      arr2[f] = arr[p1];
      p1 += 1;
    } else if p2 < arr.len() {
      arr2[f] = arr[p2];
      p2 += 1;
    } else {
      break;
    }
    f += 1;
  }

  f = 0;
  loop {
    arr[f] = arr2[f];
    f += 1;
    if f == arr.len() { break; }
  }
}

fn main() {
  let mut arr = [5,9,1,3,4,6,6,3,2];
  mergesort(&mut arr);
  println!("{:?}", arr);
}

File: mergesort.zig
const std = @import("std");

pub fn main() !void {
    var a = [_]i32{ 5, 9, 1, 3, 4, 6, 6, 3, 2 };
    try mergesort(a[0..]);
    printI32Slice(a[0..]);
}

fn mergesort(a: []i32) !void {
    if (a.len <= 1) return;
    const allocator = std.heap.page_allocator;
    var left = try allocator.alloc(i32, a.len / 2);
    defer allocator.free(left);
    var i: usize = 0;
    for (left) |*num| {
        num.* = a[i];
        i += 1;
    }
    var right = try allocator.alloc(i32, a.len - a.len / 2);
    defer allocator.free(right);
    for (right) |*num| {
        num.* = a[i];
        i += 1;
    }
    try mergesort(left);
    try mergesort(right);
    var j: usize = 0;
    var k: usize = 0;
    for (a) |*num| {
        if (j >= left.len) {
            num.* = right[k];
            k += 1;
        } else if (k >= right.len) {
            num.* = left[j];
            j += 1;
        } else if (left[j] < right[k]) {
            num.* = left[j];
            j += 1;
        } else {
            num.* = right[k];
            k += 1;
        }
    }
}

fn printI32Slice(a: []const i32) void {
    for (a) |num| {
        std.debug.print("{} ", .{num});
    }
}
